Add package: gchp

WilliamDowns opened the pull request.

WilliamDowns committed:
gchp: added new package gchp

alalazo commented on var/spack/repos/builtin/packages/gchp/package.py:
```diff
@@ -0,0 +1,68 @@
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+# ----------------------------------------------------------------------------
+# If you submit this package back to Spack as a pull request,
+# please first remove this boilerplate and all FIXME comments.
+#
+# This is a template package file for Spack.  We've put "FIXME"
+# next to all the things you'll want to change. Once you've handled
+# them, you can save this file and test your package like this:
+#
+#     spack install gchp
+#
+# You can edit this file again by typing:
+#
+#     spack edit gchp
+#
+# See the Spack documentation for more information on packaging.
+# ----------------------------------------------------------------------------
```
Please remove boilerplate comments

alalazo commented on var/spack/repos/builtin/packages/gchp/package.py:
```diff
@@ -0,0 +1,68 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
```

alalazo commented on var/spack/repos/builtin/packages/gchp/package.py:
```diff
@@ -0,0 +1,68 @@
+    variant('real8',      default=True,  description="REAL*8 precision")
+    variant('apm',        default=False, description="APM Microphysics (Experimental)")
+    variant('rrtmg',      default=False, description="RRTMG radiative transfer model")
+    variant('luo_wetdep', default=False, description="Luo et al 2019 wet deposition scheme")
```
I'd be for avoiding underscores in variant names:
```suggestion
    variant('wetdep', default=False, description="Luo et al 2019 wet deposition scheme")
```

alalazo commented on var/spack/repos/builtin/packages/gchp/package.py:
```diff
@@ -0,0 +1,68 @@
+    variant('tomas',      default=False, description="TOMAS Microphysics (Experimental)")
+
+    def cmake_args(self):
+        args = ["-DRUNDIR=" + self.prefix,
```
```suggestion
        args = [self.define("RUNDIR", self.prefix),
```

alalazo commented on var/spack/repos/builtin/packages/gchp/package.py:
```diff
@@ -0,0 +1,68 @@
+        return args
+
+    def install(self, spec, prefix):
+        with working_dir(self.build_directory):
+            inspect.getmodule(self).make(*self.install_targets)
```
Can you use `super` here instead of duplicating code from the base class?

alalazo commented on var/spack/repos/builtin/packages/gchp/package.py:
```diff
@@ -0,0 +1,68 @@
+    def install(self, spec, prefix):
+        with working_dir(self.build_directory):
+            inspect.getmodule(self).make(*self.install_targets)
+        # preserve source code structure for run directory creation
+        # can be shortened later
```
Can you add, for future reference, a bit more on why the source code is needed in the prefix?

WilliamDowns replied:
I've added some additional clarification on this portion. Thank you for your help.

alalazo reviewed (changes requested):
I left a few comments, let me know if anything is unclear.

WilliamDowns committed:
gchp: resolve reviewer edit requests

Changes include:
- Removing boilerplate header
- Updating license date
- Changing name of luo_wetdep variant to wetdep to avoid underscores in variant names
- Cleaning up argument code
- Using super in install() instead of duplicating code from base class
- Expanding on reasoning for including source code post-install

WilliamDowns committed:
gchp: remove unused import

WilliamDowns committed:
gchp: remove trailing whitespace

alalazo commented on var/spack/repos/builtin/packages/gchp/package.py:
```diff
@@ -0,0 +1,56 @@
+
+    def install(self, spec, prefix):
+        super().install(spec, prefix)
+        '''
+        Preserve source code in prefix for two reasons:
+        1. Run directory creation occurs independently of code compilation,
+        possibly multiple times depending on user needs,
+        and requires the preservation of some of the source code structure.
+        2. Run configuration is relatively complex and can result in error
+        messages that point to specific modules / lines of the source code.
+        Including source code thus facilitates runtime debugging.
+        '''
```
```suggestion
        # Preserve source code in prefix for two reasons:
        # 1. Run directory creation occurs independently of code compilation,
        # possibly multiple times depending on user needs,
        # and requires the preservation of some of the source code structure.
        # 2. Run configuration is relatively complex and can result in error
        # messages that point to specific modules / lines of the source code.
        # Including source code thus facilitates runtime debugging.
```

alalazo reviewed (changes requested):
Only a minor request to use comments instead of multi-line strings

WilliamDowns committed:
gchp: change string to series of comments

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>
